File: frontend/src/services/players.ts
import { apiClient } from "./api";
import type { Player, PlayerCreate, PlayerUpdate } from "../types";

// Create a new player
export const createPlayer = async (data: PlayerCreate): Promise<Player> => {
  const response = await apiClient.post<Player>("/players", data);
  return response.data;
};

// Get player by ID
export const getPlayer = async (playerId: number): Promise<Player> => {
  const response = await apiClient.get<Player>(`/players/${playerId}`);
  return response.data;
};

// Update player
export const updatePlayer = async (
  playerId: number,
  data: PlayerUpdate
): Promise<Player> => {
  const response = await apiClient.put<Player>(`/players/${playerId}`, data);
  return response.data;
};

// Delete player
export const deletePlayer = async (playerId: number): Promise<void> => {
  await apiClient.delete(`/players/${playerId}`);
};

File: backend/app/routers/players.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app import schemas, crud
from app.auth.dependencies import require_team_member
from app.database import get_db
from app.statistics_invalidation import (
    StatisticsCacheInvalidationReason,
    invalidate_statistics_cache,
)

router = APIRouter(
    prefix="/players",
    tags=["players"],
    dependencies=[Depends(require_team_member)],
)


@router.post("", response_model=schemas.Player, status_code=201)
def create_player(player: schemas.PlayerCreate, db: Session = Depends(get_db)):
    # Verify team exists
    team = crud.get_team(db, player.team_id)
    if not team:
        raise HTTPException(status_code=404, detail="Team not found")
    created_player = crud.create_player(db, player)
    invalidate_statistics_cache(StatisticsCacheInvalidationReason.PLAYER_CREATED)
    return created_player


@router.get("/{player_id}", response_model=schemas.Player)
def get_player(player_id: int, db: Session = Depends(get_db)):
    player = crud.get_player(db, player_id)
    if not player:
        raise HTTPException(status_code=404, detail="Player not found")
    return player


@router.put("/{player_id}", response_model=schemas.Player)
def update_player(player_id: int, player_update: schemas.PlayerUpdate, db: Session = Depends(get_db)):
    player = crud.update_player(db, player_id, player_update)
    if not player:
        raise HTTPException(status_code=404, detail="Player not found")
    invalidate_statistics_cache(StatisticsCacheInvalidationReason.PLAYER_UPDATED)
    return player


@router.delete("/{player_id}", status_code=204)
def delete_player(player_id: int, db: Session = Depends(get_db)):
    try:
        success = crud.delete_player(db, player_id)
        if not success:
            raise HTTPException(status_code=404, detail="Player not found")
        invalidate_statistics_cache(StatisticsCacheInvalidationReason.PLAYER_DELETED)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
